src/ImageSharp.Web/FormatUtilities.cs

```diff
@@ -50,21 +52,39 @@ public FormatUtilities(IOptions<ImageSharpMiddlewareOptions> options)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public string GetExtensionFromUri(string uri)
         {
-            // TODO: Investigate using span to reduce allocations here.
-            string[] parts = uri.Split('?');
-            if (parts.Length > 1 && QueryHelpers.ParseQuery(parts[1]).TryGetValue(FormatWebProcessor.Format, out StringValues ext))
+#if !NETCOREAPP3_1_OR_GREATER
+            uri = uri.ToLowerInvariant();
+#endif
+
+            int query = uri.IndexOf('?');
+            ReadOnlySpan<char> path;
+
+            if (query > -1)
+            {
+                if (QueryHelpers.ParseQuery(uri.Substring(query)).TryGetValue(FormatWebProcessor.Format, out StringValues ext))
+                {
+                    return ext;
+                }
+
+                path = uri.AsSpan(0, query);
+            }
+            else
             {
-                return ext;
+                path = uri;
             }
 
-            string path = parts[0];
             string extension = null;
             int index = 0;
             for (int i = 0; i < this.imageFormats.Length; i++)
             {
-                for (int j = 0; j < this.fileExtensions[this.imageFormats[i]].Length; j++)
+                string[] extensions = this.dottedFileExtensions[this.imageFormats[i]];
+                for (int j = 0; j < extensions.Length; j++)
                 {
-                    int li = path.LastIndexOf($".{this.fileExtensions[this.imageFormats[i]][j]}", StringComparison.OrdinalIgnoreCase);
+#if NETCOREAPP3_1_OR_GREATER
+                    int li = path.LastIndexOf(extensions[j], StringComparison.OrdinalIgnoreCase);
+#else
+                    int li = path.LastIndexOf(extensions[j]);
```

Comment:
Instead of checking for every possible match, why isn't it extracting the actual extension, and then doing a single lookup over a dictionary?

Reply:
I think this is done to make it more resistant to files with additional extensions (.png.tmp for example), although I'm not sure how often that happens, and it's not ideal for performance. If you just extract the extension and use a lookup for it:

|    NewV4 | 9.814 us | 0.0426 us | 0.0398 us | 0.4730 |      4 KB |

But I think we need to know first ( @JimBobSquarePants ) why it was done like this 🙂

Reply:
Although if you want to retain that behavior you can loop through extensions from the end, until you find one that's supported.  Takes a small hit though.

| Method |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
|------- |---------:|---------:|---------:|-------:|----------:|
|    NewV5 | 12.13 us | 0.091 us | 0.085 us | 0.7477 |      6 KB |

Reply:
We check for the '.' prefix as that excludes any filenames containing strings that match extension names. e.g. `mybmp.jpg`

Reply:
OK, then it's not needed to check for multiple extensions (which seems rather pointless anyway), so the NewV4 variant can be used (with 1KB less allocation on .NET Core 3.1+), which would further cut down the time ~6x. Let me commit that version.

Reply:
Committed, tests still pass 🙂

Reply:
If you introduce the new `dottedFileExtensions` dictionary (_no need to use anything other than the default comparer_) as you have and use a string for the key for `fileExtensions` the entire method can become as simple as this.

```c#
public string GetExtensionFromUri(string uri)
{
    int queryIndex = uri.IndexOf('?');
    if (queryIndex != -1 && QueryHelpers.ParseQuery(uri.Substring(queryIndex)).TryGetValue(FormatWebProcessor.Format, out StringValues ext))
    {
        return ext;
    }

    string extension = null;
    int index = 0;
    for (int i = 0; i < this.imageFormats.Length; i++)
    {
        string name = this.imageFormats[i].Name;
        string[] extensions = this.fileExtensions[name];
        string[] dottedExtensions = this.dottedFileExtensions[name];
        for (int j = 0; j < dottedExtensions.Length; j++)
        {
            int li = uri.LastIndexOf(dottedExtensions[j], StringComparison.OrdinalIgnoreCase);
            if (li < index)
            {
                continue;
            }

            index = li;
            extension = extensions[j];
        }
    }

    return extension;
}
```

Reply:
>OK, then it's not needed to check for multiple extensions (which seems rather pointless anyway)

No you still need to check for both scenarios. Someone could appear with `mybmp.jpg` and `mybmp.png.jpg`

Reply:
> 
> 
> > OK, then it's not needed to check for multiple extensions (which seems rather pointless anyway)
> 
> No you still need to check for both scenarios. Someone could appear with `mybmp.jpg` and `mybmp.png.jpg`

What would you expect the format to be in this case? png or jpg? I would expect jpg, which would be valid with the code as-is.

The code above is much slower than simply extracting the extension.

Reply:
BTW, if you want to loop through everything, it's trivial to change the if to a while and add ```path = path.Slice(0, extensionIndex);``` at the end of the while. Then it will loop from end to start.

Reply:
I'm not sure using the hashcode as you have is safe here. 
https://stackoverflow.com/a/7969189/427899

Reply:
It's just over the extension and I've checked, there are no collisions there. The post is from 2011 and in .NET Core the hashcode algorithm changed to Marvan and no longer collides so much :)

Reply:
What's the cost of `Substring` nowadays? Has that been optimized?

Reply:
Considerably, it's the difference between NewV6 and NewV2.

Reply:
Ah no, I don't mean a collision between extensions I meant a collision between the hashed sliced input value and a proper extension. The probability of getting a string with a dot and following value that collided would be crazy small though! :)

I'm wondering if we can avoid the dictionary? I have an Idea I'd like to run past you?

Let's say that instead of using a dictionary `fileExtensions` was instead an array containing of all the combined extensions from all image formats. Stealing a lot of the clever stuff you've done already we could slice the input uri and do a lookup against that array. `ToString()` on a `ReadonlySpan<char>` costs nothing and the ordinal comparison should be fast enough. What do you think?

```c#
public string GetExtensionFromUri(string uri)
{
    ReadOnlySpan<char> path;
    int queryIndex = uri.IndexOf('?');
    if (queryIndex != -1)
    {
        if (QueryHelpers.ParseQuery(uri.Substring(queryIndex)).TryGetValue(FormatWebProcessor.Format, out StringValues ext))
        {
            return ext;
        }

        path = uri.AsSpan(0, queryIndex);
    }
    else
    {
        path = uri;
    }

    int index;
    if ((index = path.LastIndexOf('.')) != -1)
    {
        uri = path.Slice(index + 1).ToString();
        string[] extensions = this.fileExtensions;
        for (int i = 0; i < extensions.Length; i++)
        {
            string extension = extensions[i];
            if (extension.Equals(uri, StringComparison.OrdinalIgnoreCase))
            {
                return extension;
            }
        }
    }

    return null;
}
```

Reply:
Please don't index on a hashcode, it doesn't make sense (to me), both for correctness and performance.

When you think an iteration is faster than a lookup, use an `HybridDictionary`, it does it for you, converts to an actual dictionary if there are many elements.

Also the `.ToString()` "will" allocate, obviously. You can do the same comparison allocation-free by only using slices and span comparisons on the uri (endswith) if you really don't want to use a dictionary. Then I would suggest to order the extensions by MRU since it's likely that the same extensions are used over and over. And if not MRU as it might get costly, have two collections, "used" and "never-used", then it might be fine iterating on the "used" collections, and doing a lookup on the "never-used" as a fallback.

All this is very fun, more fun that the actual gains ;)

Reply:
Thanks @sebastienros. I have a page with a lot of images on it and this takes the majority of the page load time. Not so noticeable on my dev machine, but on a slow machine it takes quite some time as this gets called per image request.
@JimBobSquarePants 
I have changed the implementation to use a List and loop through it, this would still be ~6x as fast as the original implementation.

| Method |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
|------- |---------:|---------:|---------:|-------:|----------:|
|    New | 11.36 us | 0.206 us | 0.182 us | 0.3967 |      3 KB |

Reply:
> I have a page with a lot of images on it

What I meant is the number of distinct extensions, even if you have thousands of images on this page, they probably are using the same extension. I would also assume most extensions used in an app are always the same, like `.png`, `.jpg`, ... so splitting the list of extensions based on usage might be a better optimization since the main list would then have just a few extensions in it, and a second list have the remaining ones.

Reply:
They'll get loaded in the order as defined in the `Configuration` class. Current order is png, jpg, gif, bmp, tga. Someone can define their own configuration instance via `ImageSharpMiddlewareOptions` if need be. 

Can't believe I didn't know about the `ROS<char>` equality comparison. Super useful!

Reply:
I think it would probably be the best to keep order the same as the one in the config, otherwise it would be another thing to configure. The order of png, jpg seems like the good default based on web usage. With webp maybe it can be added before png (or after).

Reply:
I'm actually gonna PR upstream into the 2.0 alphas to change the registration order to match most common usage.